concrete_biopython/FheSeq.py

```diff
@@ -1,14 +1,200 @@
 from abc import ABC
 from concrete import fhe
 import numpy as np
-import numbers
+import numbers, itertools
 
-from concrete_biopython.SeqWrapper import SeqWrapper
+from Bio.Seq import Seq, MutableSeq
 
 Tracer = fhe.tracing.tracer.Tracer
 
 _NOT_IMPLEMENTED_FHE =  Exception('This function is not compatible with FHE')
 
+def boolean_multiplication(value, boolean):
+    return fhe.univariate(lambda packing: np.where(packing % 2, packing // 2, 0))(value * 2 + boolean)
+
+
+class Alphabets():
+    DNA_RNA = "\0ACGTU"
+    PROTEINS = "\0*ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    ASCII = '\0!\"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\t\n'
+    LOWERCASE = "\0abcdefghijklmnopqrstuvwxyz"
+    UPPERCASE = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    LETTERS = "\0ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+    DIGITS = "\0"+"0123456789"
```

Comment:
What was the purpose of `\0` again?

Reply:
I need an empty character to deal with the case where two sequences don't have the same length, so I can extend one to match the length of the other and not change the content

Reply:
Why do you need to do that, couldn't it be different across operands?

Reply:
Yes it would be useful! I just wanted to see how it's used exactly. Like are you padding every input to the biggest len for example?

Reply:
you can check the function `__lt__` for instance:
```python
 def __lt__(self, other):
        """Computes whether a sequence is lower than another, in alphabetical order

        A sequence A is lower than a sequence B if and only if:
        there exist an index i such that: A[i] < B[i]  AND  for all k<i, A[k] <= B[k]

        If sequence have different length, the shorter one is considered to have extra empty characters
        where the empty character is the lowest in alphabetical order
        """
        # prepare arrays if they have different size
        difflen = len(other) - len(self)

        # append a zero to the shortest sequence and crop the other one
        # zero is always <= to any other character so no need to append more zeros
        if difflen > 0:
            A = np.concatenate((self._data, fhe.zeros(1).reshape(1)), axis=0)
            B = other._data[0 : A.size]
        elif difflen < 0:
            B = np.concatenate((other._data, fhe.zeros(1).reshape(1)), axis=0)
            A = self._data[0 : B.size]
        else:
            A = self._data[:]  # copy array
            B = other._data[:]  # copy array

        n = A.size

        if n == 0:
            return fhe.one()  # special case if both arrays are empty

        # Fast computation of (A < B)
        # (This algorithm is inspired from the behavior of the failure of subtraction(A,B) algorithm when A < B)
        borrow = 0
        for i in range(n):
            # report borrow
            borrow = A[-i - 1] - B[-i - 1] - borrow < 0

        # When A < B, subtracting B from A fails and overflows, we use the overflow value borrow to know if A < B
        return borrow
```
coparing the sequence "AB" with "ABCDE" will be equivalent to compare "AB\0" with "ABC" and the function will tell that the first one is "lower than" the other because 0 is smaller than any other values, lower than C here